server/controllers/CollaboratorController.js

```diff
@@ -0,0 +1,105 @@
+import User from './../models/User';
+import Collaborator from './../models/Collaborator';
+
+import Search from './../helpers/Search';
+
+/**
+ * @class CollaboratorController
+ */
+class CollaboratorController {
+  /**
+   * Gets all collaborators on a to-do
+   *
+   * @static
+   *
+   * @param {Object} request - request object
+   * @param {Object} response - response object
+   *
+   * @memberof CollaboratorController
+   *
+   * @returns {void}
+   */
+  static async getCollaborators(request, response) {
+    const { toDoId } = request.params;
+
+    const collaborators = await Search.searchAll(Collaborator, { toDoId });
+
+    const refinedCollaborators = collaborators.map(collaborator => ({
+      toDoId: collaborator.toDoId,
+      collaborator: collaborator.collaboratorId,
+      readOnly: collaborator.readOnly,
+    }));
+
+    response.status(200).json({
+      collaborators: refinedCollaborators,
+    });
+  }
+
+  /**
+   * Add collaborators to a to-do
+   *
+   * @static
+   *
+   * @param {Object} request - request object
+   * @param {Object} response - request object
+   *
+   * @memberof CollaboratorController
+   *
+   * @returns {void}
+   */
+  static async addCollaborators(request, response) {
+    request.checkBody('email', 'Invalid email').isEmail();
+    request.checkBody('readOnly', 'Permission must be a boolean').isBoolean();
+
+    const requestErrors = request.validationErrors();
+
+    if (requestErrors) {
+      response.status(400).json({
+        errors: requestErrors,
+      });
+    } else {
+      request.sanitizeBody('email').escape();
+
+      const { toDoId } = request.params;
+      const { email } = request.body;
+      const existingUser = await Search.searchOne(User, { email });
+
+      if (!existingUser) {
+        response.status(400).json({
+          error: 'User does not exist',
+        });
+      } else {
+        const collaboratorId = existingUser.id;
+
+        const existingCollaborator = await Search.searchOne(
+          Collaborator,
+          {
+            collaboratorId,
+            toDoId
+          }
+        );
+
+        if (existingCollaborator) {
+          response.status(409).json({
+            error: 'Collaborator is already added',
+          });
+        } else {
+          const collaborator = await Collaborator({ toDoId, collaboratorId })
+            .save();
+
+          const refinedCollaborator = {
```

Comment:
The reason why I am refining the data here is because MongoDB adds extra properties in addition to the ones I defined in my models. That is why I am refining, to grab just the properties I need. I.e., the properties I defined myself

Reply:
Another reason why I am refining is because the unique ObjectId generated by MongoDB has the key `_id`, which isn't descriptive enough. So when I refine, I would rename it to `userId`, for example

Reply:
Oh, I see. Ok 👍